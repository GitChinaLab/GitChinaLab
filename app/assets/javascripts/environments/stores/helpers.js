/**
 * @param {Object} environment
 * @returns {Object}
 */
export const setDeployBoard = (oldEnvironmentState, environment) => {
  let parsedEnvironment = environment;
  if (!environment.isFolder && environment.rollout_status) {
    parsedEnvironment = {
      ...environment,
      hasDeployBoard: true,
      isDeployBoardVisible:
        oldEnvironmentState.isDeployBoardVisible === false
          ? oldEnvironmentState.isDeployBoardVisible
          : true,
      deployBoardData:
        environment.rollout_status.status === 'found' ? environment.rollout_status : {},
      isLoadingDeployBoard: environment.rollout_status.status === 'loading',
      isEmptyDeployBoard: environment.rollout_status.status === 'not_found',
    };
  }
  return parsedEnvironment;
};
