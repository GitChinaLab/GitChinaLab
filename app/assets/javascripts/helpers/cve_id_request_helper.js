export function createCveIdRequestIssueBody(fullPath, iid) {
  return `### Vulnerability Submission

**NOTE:** Only maintainers of GitLab-hosted projects may request a CVE for
a vulnerability within their project.

Project issue: ${fullPath}#${iid}

#### Publishing Schedule

After a CVE request is validated, a CVE identifier will be assigned. On what
schedule should the details of the CVE be published?

* [ ] Publish immediately
* [ ] Wait to publish

<!--
Please fill out the yaml codeblock below
-->

\`\`\`yaml
reporter:
  name: "TODO" # "First Last"
  email: "TODO" # "email@domain.tld"
vulnerability:
  description: "TODO" # "[VULNTYPE] in [COMPONENT] in [VENDOR][PRODUCT] [VERSION] allows [ATTACKER] to [IMPACT] via [VECTOR]"
  cwe: "TODO" # "CWE-22" # Path Traversal
  product:
    gitlab_path: "${fullPath}"
    vendor: "TODO" # "Deluxe Sandwich Maker Company"
    name: "TODO" # "Deluxe Sandwich Maker 2"
    affected_versions:
      - "TODO" # "1.2.3"
      - "TODO" # ">1.3.0, <=1.3.9"
    fixed_versions:
      - "TODO" # "1.2.4"
      - "TODO" # "1.3.10"
  impact: "TODO" # "CVSS v3 string" # https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator
  solution: "TODO" # "Upgrade to version 1.2.4 or 1.3.10"
  credit: "TODO"
  references:
    - "TODO" # "https://some.domain.tld/a/reference"
\`\`\`

CVSS scores can be computed by means of the [NVD CVSS Calculator](https://nvd.nist.gov/vuln-metrics/cvss/v3-calculator).

/relate ${fullPath}#${iid}
/label ~"devops::secure" ~"group::vulnerability research" ~"vulnerability research::cve" ~"advisory::queued"
  `;
}
