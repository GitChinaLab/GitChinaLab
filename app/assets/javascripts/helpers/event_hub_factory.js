/**
 * An event hub with a Vue instance like API
 *
 * NOTE: There's an [issue open][4] to eventually remove this when some
 * coupling in our codebase has been fixed.
 *
 * NOTE: This is a derivative work from [mitt][1] v1.2.0 which is licensed by
 * [MIT License][2] Â© [Jason Miller][3]
 *
 * [1]: https://github.com/developit/mitt
 * [2]: https://opensource.org/licenses/MIT
 * [3]: https://jasonformat.com/
 * [4]: https://gitlab.com/gitlab-org/gitlab/-/issues/223864
 */
class EventHub {
  constructor() {
    this.$_all = new Map();
  }

  dispose() {
    this.$_all.clear();
  }

  /**
   * Register an event handler for the given type.
   *
   * @param {string|symbol} type Type of event to listen for
   * @param {Function} handler Function to call in response to given event
   */
  $on(type, handler) {
    const handlers = this.$_all.get(type);
    const added = handlers && handlers.push(handler);

    if (!added) {
      this.$_all.set(type, [handler]);
    }
  }

  /**
   * Remove an event handler or all handlers for the given type.
   *
   * @param {string|symbol} type Type of event to unregister `handler`
   * @param {Function} handler Handler function to remove
   */
  $off(type, handler) {
    const handlers = this.$_all.get(type) || [];

    const newHandlers = handler ? handlers.filter((x) => x !== handler) : [];

    if (newHandlers.length) {
      this.$_all.set(type, newHandlers);
    } else {
      this.$_all.delete(type);
    }
  }

  /**
   * Add an event listener to type but only trigger it once
   *
   * @param {string|symbol} type Type of event to listen for
   * @param {Function} handler Handler function to call in response to event
   */
  $once(type, handler) {
    const wrapHandler = (...args) => {
      this.$off(type, wrapHandler);
      handler(...args);
    };
    this.$on(type, wrapHandler);
  }

  /**
   * Invoke all handlers for the given type.
   *
   * @param {string|symbol} type The event type to invoke
   * @param {Any} [evt] Any value passed to each handler
   */
  $emit(type, ...args) {
    const handlers = this.$_all.get(type) || [];

    handlers.forEach((handler) => {
      handler(...args);
    });
  }
}

/**
 * Return a Vue like event hub
 *
 * - $on
 * - $off
 * - $once
 * - $emit
 *
 * Please note, this was once implemented with `mitt`, but since then has been reverted
 * because of some API issues. https://gitlab.com/gitlab-org/gitlab/-/merge_requests/35074
 *
 * We'd like to shy away from using a full fledged Vue instance from this in the future.
 */
export default () => {
  return new EventHub();
};
