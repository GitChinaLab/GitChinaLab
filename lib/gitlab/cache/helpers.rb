# frozen_string_literal: true

module Gitlab
  module Cache
    module Helpers
      # @return [ActiveSupport::Duration]
      DEFAULT_EXPIRY = 1.day

      # @return [ActiveSupport::Cache::Store]
      def cache
        Rails.cache
      end

      def render_cached(obj_or_collection, with:, cache_context: -> (_) { current_user&.cache_key }, expires_in: Gitlab::Cache::Helpers::DEFAULT_EXPIRY, **presenter_args)
        json =
          if obj_or_collection.is_a?(Enumerable)
            cached_collection(
              obj_or_collection,
              presenter: with,
              presenter_args: presenter_args,
              context: cache_context,
              expires_in: expires_in
            )
          else
            cached_object(
              obj_or_collection,
              presenter: with,
              presenter_args: presenter_args,
              context: cache_context,
              expires_in: expires_in
            )
          end

        render Gitlab::Json::PrecompiledJson.new(json)
      end

      private

      # Optionally uses a `Proc` to add context to a cache key
      #
      # @param object [Object] must respond to #cache_key
      # @param context [Proc] a proc that will be called with the object as an argument, and which should return a
      #                       string or array of strings to be combined into the cache key
      # @return [String]
      def contextual_cache_key(presenter, object, context)
        return object.cache_key if context.nil?

        [presenter.class.name, object.cache_key, context.call(object)].flatten.join(":")
      end

      # Used for fetching or rendering a single object
      #
      # @param object [Object] the object to render
      # @param presenter [Grape::Entity]
      # @param presenter_args [Hash] keyword arguments to be passed to the entity
      # @param context [Proc]
      # @param expires_in [ActiveSupport::Duration, Integer] an expiry time for the cache entry
      # @return [String]
      def cached_object(object, presenter:, presenter_args:, context:, expires_in:)
        cache.fetch(contextual_cache_key(presenter, object, context), expires_in: expires_in) do
          Gitlab::Json.dump(presenter.represent(object, **presenter_args).as_json)
        end
      end

      # Used for fetching or rendering multiple objects
      #
      # @param objects [Enumerable<Object>] the objects to render
      # @param presenter [Grape::Entity]
      # @param presenter_args [Hash] keyword arguments to be passed to the entity
      # @param context [Proc]
      # @param expires_in [ActiveSupport::Duration, Integer] an expiry time for the cache entry
      # @return [Array<String>]
      def cached_collection(collection, presenter:, presenter_args:, context:, expires_in:)
        json = fetch_multi(presenter, collection, context: context, expires_in: expires_in) do |obj|
          Gitlab::Json.dump(presenter.represent(obj, **presenter_args).as_json)
        end

        json.values
      end

      # An adapted version of ActiveSupport::Cache::Store#fetch_multi.
      #
      # The original method only provides the missing key to the block,
      # not the missing object, so we have to create a map of cache keys
      # to the objects to allow us to pass the object to the missing value
      # block.
      #
      # The result is that this is functionally identical to `#fetch`.
      def fetch_multi(presenter, *objs, context:, **kwargs)
        objs.flatten!
        map = multi_key_map(presenter, objs, context: context)

        # TODO: `contextual_cache_key` should be constructed based on the guideline https://docs.gitlab.com/ee/development/redis.html#multi-key-commands.
        Gitlab::Instrumentation::RedisClusterValidator.allow_cross_slot_commands do
          cache.fetch_multi(*map.keys, **kwargs) do |key|
            yield map[key]
          end
        end
      end

      # @param objects [Enumerable<Object>] objects which _must_ respond to `#cache_key`
      # @param context [Proc] a proc that can be called to help generate each cache key
      # @return [Hash]
      def multi_key_map(presenter, objects, context:)
        objects.index_by do |object|
          contextual_cache_key(presenter, object, context)
        end
      end
    end
  end
end
