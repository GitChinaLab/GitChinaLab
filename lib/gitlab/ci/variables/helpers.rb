# frozen_string_literal: true

module Gitlab
  module Ci
    module Variables
      module Helpers
        class << self
          def merge_variables(current_vars, new_vars)
            current_vars = transform_from_yaml_variables(current_vars)
            new_vars = transform_from_yaml_variables(new_vars)

            transform_to_yaml_variables(
              current_vars.merge(new_vars)
            )
          end

          def transform_to_yaml_variables(vars)
            vars.to_h.map do |key, value|
              { key: key.to_s, value: value, public: true }
            end
          end

          def transform_from_yaml_variables(vars)
            return vars.stringify_keys if vars.is_a?(Hash)

            vars.to_a.to_h { |var| [var[:key].to_s, var[:value]] }
          end

          def inherit_yaml_variables(from:, to:, inheritance:)
            merge_variables(apply_inheritance(from, inheritance), to)
          end

          private

          def apply_inheritance(variables, inheritance)
            case inheritance
            when true then variables
            when false then {}
            when Array then variables.select { |var| inheritance.include?(var[:key]) }
            end
          end
        end
      end
    end
  end
end
