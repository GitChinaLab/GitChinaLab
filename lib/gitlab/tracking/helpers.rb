# frozen_string_literal: true

module Gitlab
  module Tracking
    module Helpers
      def dnt_enabled?
        Gitlab::Utils.to_boolean(request.headers['DNT'])
      end

      def trackable_html_request?
        request.format.html? && !dnt_enabled?
      end
    end
  end
end
