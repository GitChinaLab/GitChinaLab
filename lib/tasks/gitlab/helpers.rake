# frozen_string_literal: true

# Prevent StateMachine warnings from outputting during a cron task
StateMachines::Machine.ignore_method_conflicts = true if ENV['CRON']

task :gitlab_environment do
  Rake::Task[:environment].invoke unless ENV['SKIP_RAILS_ENV_IN_RAKE']

  extend SystemCheck::Helpers
end
