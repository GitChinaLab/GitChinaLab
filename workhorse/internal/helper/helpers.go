package helper

import (
	"bytes"
	"errors"
	"io/ioutil"
	"mime"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"syscall"

	"github.com/sebest/xff"
	"gitlab.com/gitlab-org/labkit/log"
	"gitlab.com/gitlab-org/labkit/mask"
)

const NginxResponseBufferHeader = "X-Accel-Buffering"

func logErrorWithFields(r *http.Request, err error, fields log.Fields) {
	if err != nil {
		CaptureRavenError(r, err, fields)
	}

	printError(r, err, fields)
}

func CaptureAndFail(w http.ResponseWriter, r *http.Request, err error, msg string, code int) {
	http.Error(w, msg, code)
	logErrorWithFields(r, err, nil)
}

func Fail500(w http.ResponseWriter, r *http.Request, err error) {
	CaptureAndFail(w, r, err, "Internal server error", http.StatusInternalServerError)
}

func Fail500WithFields(w http.ResponseWriter, r *http.Request, err error, fields log.Fields) {
	http.Error(w, "Internal server error", http.StatusInternalServerError)
	logErrorWithFields(r, err, fields)
}

func RequestEntityTooLarge(w http.ResponseWriter, r *http.Request, err error) {
	CaptureAndFail(w, r, err, "Request Entity Too Large", http.StatusRequestEntityTooLarge)
}

func printError(r *http.Request, err error, fields log.Fields) {
	if r != nil {
		entry := log.WithContextFields(r.Context(), log.Fields{
			"method": r.Method,
			"uri":    mask.URL(r.RequestURI),
		})
		entry.WithFields(fields).WithError(err).Error()
	} else {
		log.WithFields(fields).WithError(err).Error("unknown error")
	}
}

func SetNoCacheHeaders(header http.Header) {
	header.Set("Cache-Control", "no-cache, no-store, max-age=0, must-revalidate")
	header.Set("Pragma", "no-cache")
	header.Set("Expires", "Fri, 01 Jan 1990 00:00:00 GMT")
}

func OpenFile(path string) (file *os.File, fi os.FileInfo, err error) {
	file, err = os.Open(path)
	if err != nil {
		return
	}

	defer func() {
		if err != nil {
			file.Close()
		}
	}()

	fi, err = file.Stat()
	if err != nil {
		return
	}

	// The os.Open can also open directories
	if fi.IsDir() {
		err = &os.PathError{
			Op:   "open",
			Path: path,
			Err:  errors.New("path is directory"),
		}
		return
	}

	return
}

func URLMustParse(s string) *url.URL {
	u, err := url.Parse(s)
	if err != nil {
		log.WithError(err).WithField("url", s).Fatal("urlMustParse")
	}
	return u
}

func HTTPError(w http.ResponseWriter, r *http.Request, error string, code int) {
	if r.ProtoAtLeast(1, 1) {
		// Force client to disconnect if we render request error
		w.Header().Set("Connection", "close")
	}

	http.Error(w, error, code)
}

func HeaderClone(h http.Header) http.Header {
	h2 := make(http.Header, len(h))
	for k, vv := range h {
		vv2 := make([]string, len(vv))
		copy(vv2, vv)
		h2[k] = vv2
	}
	return h2
}

func CleanUpProcessGroup(cmd *exec.Cmd) {
	if cmd == nil {
		return
	}

	process := cmd.Process
	if process != nil && process.Pid > 0 {
		// Send SIGTERM to the process group of cmd
		syscall.Kill(-process.Pid, syscall.SIGTERM)
	}

	// reap our child process
	cmd.Wait()
}

func ExitStatus(err error) (int, bool) {
	exitError, ok := err.(*exec.ExitError)
	if !ok {
		return 0, false
	}

	waitStatus, ok := exitError.Sys().(syscall.WaitStatus)
	if !ok {
		return 0, false
	}

	return waitStatus.ExitStatus(), true
}

func DisableResponseBuffering(w http.ResponseWriter) {
	w.Header().Set(NginxResponseBufferHeader, "no")
}

func AllowResponseBuffering(w http.ResponseWriter) {
	w.Header().Del(NginxResponseBufferHeader)
}

func FixRemoteAddr(r *http.Request) {
	// Unix domain sockets have a remote addr of @. This will make the
	// xff package lookup the X-Forwarded-For address if available.
	if r.RemoteAddr == "@" {
		r.RemoteAddr = "127.0.0.1:0"
	}
	r.RemoteAddr = xff.GetRemoteAddr(r)
}

func SetForwardedFor(newHeaders *http.Header, originalRequest *http.Request) {
	if clientIP, _, err := net.SplitHostPort(originalRequest.RemoteAddr); err == nil {
		var header string

		// If we aren't the first proxy retain prior
		// X-Forwarded-For information as a comma+space
		// separated list and fold multiple headers into one.
		if prior, ok := originalRequest.Header["X-Forwarded-For"]; ok {
			header = strings.Join(prior, ", ") + ", " + clientIP
		} else {
			header = clientIP
		}
		newHeaders.Set("X-Forwarded-For", header)
	}
}

func IsContentType(expected, actual string) bool {
	parsed, _, err := mime.ParseMediaType(actual)
	return err == nil && parsed == expected
}

func IsApplicationJson(r *http.Request) bool {
	contentType := r.Header.Get("Content-Type")
	return IsContentType("application/json", contentType)
}

func ReadRequestBody(w http.ResponseWriter, r *http.Request, maxBodySize int64) ([]byte, error) {
	limitedBody := http.MaxBytesReader(w, r.Body, maxBodySize)
	defer limitedBody.Close()

	return ioutil.ReadAll(limitedBody)
}

func CloneRequestWithNewBody(r *http.Request, body []byte) *http.Request {
	newReq := *r
	newReq.Body = ioutil.NopCloser(bytes.NewReader(body))
	newReq.Header = HeaderClone(r.Header)
	newReq.ContentLength = int64(len(body))
	return &newReq
}
