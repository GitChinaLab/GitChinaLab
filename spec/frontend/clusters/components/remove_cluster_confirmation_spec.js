import { GlModal, GlSprintf } from '@gitlab/ui';
import { mount } from '@vue/test-utils';
import { stubComponent } from 'helpers/stub_component';
import RemoveClusterConfirmation from '~/clusters/components/remove_cluster_confirmation.vue';
import SplitButton from '~/vue_shared/components/split_button.vue';

describe('Remove cluster confirmation modal', () => {
  let wrapper;

  const createComponent = ({ props = {}, stubs = {} } = {}) => {
    wrapper = mount(RemoveClusterConfirmation, {
      propsData: {
        clusterPath: 'clusterPath',
        clusterName: 'clusterName',
        ...props,
      },
      stubs,
    });
  };

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders splitbutton with modal included', () => {
    createComponent();
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('split button dropdown', () => {
    const findModal = () => wrapper.findComponent(GlModal);
    const findSplitButton = () => wrapper.findComponent(SplitButton);

    beforeEach(() => {
      createComponent({
        props: { clusterName: 'my-test-cluster' },
        stubs: { GlSprintf, GlModal: stubComponent(GlModal) },
      });
      jest.spyOn(findModal().vm, 'show').mockReturnValue();
    });

    it('opens modal with "cleanup" option', async () => {
      findSplitButton().vm.$emit('remove-cluster-and-cleanup');

      await wrapper.vm.$nextTick();

      expect(findModal().vm.show).toHaveBeenCalled();
      expect(wrapper.vm.confirmCleanup).toEqual(true);
      expect(findModal().html()).toContain(
        '<strong>To remove your integration and resources, type <code>my-test-cluster</code> to confirm:</strong>',
      );
    });

    it('opens modal without "cleanup" option', async () => {
      findSplitButton().vm.$emit('remove-cluster');

      await wrapper.vm.$nextTick();

      expect(findModal().vm.show).toHaveBeenCalled();
      expect(wrapper.vm.confirmCleanup).toEqual(false);
      expect(findModal().html()).toContain(
        '<strong>To remove your integration, type <code>my-test-cluster</code> to confirm:</strong>',
      );
    });

    describe('with cluster management project', () => {
      beforeEach(() => {
        createComponent({ props: { hasManagementProject: true } });
      });

      it('renders regular button instead', () => {
        expect(findSplitButton().exists()).toBe(false);
        expect(wrapper.find('[data-testid="btnRemove"]').exists()).toBe(true);
      });
    });
  });
});
