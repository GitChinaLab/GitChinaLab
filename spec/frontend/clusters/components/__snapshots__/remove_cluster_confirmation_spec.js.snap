// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove cluster confirmation modal renders splitbutton with modal included 1`] = `
<div
  class="gl-display-flex gl-justify-content-end"
>
  <div
    class="dropdown b-dropdown gl-new-dropdown btn-group"
    menu-class="dropdown-menu-large"
  >
    <button
      class="btn btn-danger btn-md gl-button split-content-button"
      type="button"
    >
      <!---->
       
      <!---->
       
      <span
        class="gl-new-dropdown-button-text"
      >
        Remove integration and resources
      </span>
       
      <!---->
    </button>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn dropdown-toggle btn-danger btn-md gl-button gl-dropdown-toggle dropdown-toggle-split"
      type="button"
    >
      <span
        class="sr-only"
      >
        Toggle dropdown
      </span>
    </button>
    <ul
      class="dropdown-menu dropdown-menu-large"
      role="menu"
      tabindex="-1"
    >
      <div
        class="gl-new-dropdown-inner"
      >
        <!---->
         
        <!---->
         
        <div
          class="gl-new-dropdown-contents"
        >
          <!---->
           
          <li
            class="gl-new-dropdown-item"
            role="presentation"
          >
            <button
              class="dropdown-item"
              role="menuitem"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="gl-icon s16 gl-new-dropdown-item-check-icon gl-mt-3 gl-align-self-start"
                data-testid="dropdown-item-checkbox"
                role="img"
              >
                <use
                  href="#mobile-issue-close"
                />
              </svg>
               
              <!---->
               
              <!---->
               
              <div
                class="gl-new-dropdown-item-text-wrapper"
              >
                <p
                  class="gl-new-dropdown-item-text-primary"
                >
                  <strong>
                    Remove integration and resources
                  </strong>
                   
                  <div>
                    Deletes all GitLab resources attached to this cluster during removal
                  </div>
                </p>
                 
                <!---->
              </div>
               
              <!---->
            </button>
          </li>
           
          <li
            class="gl-new-dropdown-divider"
            role="presentation"
          >
            <hr
              aria-orientation="horizontal"
              class="dropdown-divider"
              role="separator"
            />
          </li>
          <li
            class="gl-new-dropdown-item"
            role="presentation"
          >
            <button
              class="dropdown-item"
              role="menuitem"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="gl-icon s16 gl-new-dropdown-item-check-icon gl-visibility-hidden gl-mt-3 gl-align-self-start"
                data-testid="dropdown-item-checkbox"
                role="img"
              >
                <use
                  href="#mobile-issue-close"
                />
              </svg>
               
              <!---->
               
              <!---->
               
              <div
                class="gl-new-dropdown-item-text-wrapper"
              >
                <p
                  class="gl-new-dropdown-item-text-primary"
                >
                  <strong>
                    Remove integration
                  </strong>
                   
                  <div>
                    Removes cluster from project but keeps associated resources
                  </div>
                </p>
                 
                <!---->
              </div>
               
              <!---->
            </button>
          </li>
           
          <!---->
        </div>
         
        <!---->
      </div>
    </ul>
  </div>
   
  <!---->
</div>
`;
